File: src/schemas/outbound_activity.py
# Copyright (c) 2026 SiinXu / StockPulse contributors
# SPDX-License-Identifier: AGPL-3.0-only
"""Read-only contracts for outbound HTTP policy transparency."""

from __future__ import annotations

from typing import List, Literal

from pydantic import BaseModel, ConfigDict, Field

OUTBOUND_ACTIVITY_MAX_PAGE_SIZE = 100

OutboundDecision = Literal["allowed", "blocked"]


class _StrictOutboundModel(BaseModel):
    model_config = ConfigDict(
        allow_inf_nan=False,
        extra="forbid",
        strict=True,
        str_strip_whitespace=True,
    )


class LocalOnlyModeStatus(_StrictOutboundModel):
    enabled: bool = Field(...)
    env_key: str = Field(default="LOCAL_ONLY_MODE")
    policy: str = Field(default="non_loopback_denied")
    allowed_destination_classes: List[str] = Field(default_factory=lambda: ["loopback"])
    blocked_error_reason: str = Field(default="local_only_mode_blocked")


class OutboundActivityItem(_StrictOutboundModel):
    occurred_at: str = Field(...)
    decision: OutboundDecision = Field(...)
    destination_class: str = Field(...)
    scheme: str = Field(...)
    host_type: str = Field(...)
    reason: str = Field(...)
    correlation_id: str = Field(...)
    local_only_mode: bool = Field(...)
    allowlisted: bool = Field(...)


class OutboundActivityPage(_StrictOutboundModel):
    local_only_mode: bool = Field(...)
    items: List[OutboundActivityItem] = Field(default_factory=list)
    limit: int = Field(..., ge=1, le=OUTBOUND_ACTIVITY_MAX_PAGE_SIZE)
    returned: int = Field(..., ge=0)
    max_retained: int = Field(...)

File: apps/dsa-web/src/types/outboundActivity.ts
// Copyright (c) 2026 SiinXu / StockPulse contributors
// SPDX-License-Identifier: AGPL-3.0-only

import type { components, operations } from './api.generated';

type CamelCase<S extends string> = S extends `${infer Head}_${infer Tail}`
  ? `${Head}${Capitalize<CamelCase<Tail>>}`
  : S;

type CamelizeKeys<T> = T extends readonly (infer U)[]
  ? CamelizeKeys<U>[]
  : T extends object
    ? { [K in keyof T as CamelCase<K & string>]: CamelizeKeys<T[K]> }
    : T;

type Override<T, U> = Omit<T, keyof U> & U;

type OpenApiLocalOnly = components['schemas']['LocalOnlyModeStatus'];
type OpenApiItem = components['schemas']['OutboundActivityItem'];
type OpenApiPage = components['schemas']['OutboundActivityPage'];
type OpenApiLocalOnlyGet200 =
  operations['get_local_only_mode_status_api_v1_security_local_only_get']['responses']['200']['content']['application/json'];
type OpenApiOutboundGet200 =
  operations['list_outbound_activity_api_v1_security_outbound_activity_get']['responses']['200']['content']['application/json'];

type _Assert<T extends true> = T;
type _LocalOnlyGetIsComponent = _Assert<OpenApiLocalOnlyGet200 extends OpenApiLocalOnly ? true : false>;
type _LocalOnlyComponentIsGet = _Assert<OpenApiLocalOnly extends OpenApiLocalOnlyGet200 ? true : false>;
type _OutboundGetIsComponent = _Assert<OpenApiOutboundGet200 extends OpenApiPage ? true : false>;
type _OutboundComponentIsGet = _Assert<OpenApiPage extends OpenApiOutboundGet200 ? true : false>;

type _OpenApiAnchors = [
  _LocalOnlyGetIsComponent,
  _LocalOnlyComponentIsGet,
  _OutboundGetIsComponent,
  _OutboundComponentIsGet,
];
type _BindOpenApiAnchors<T> = [_OpenApiAnchors] extends [unknown] ? T : T;

export type OutboundDecision = 'allowed' | 'blocked';

export type LocalOnlyModeStatus = _BindOpenApiAnchors<Override<CamelizeKeys<OpenApiLocalOnly>, {
  enabled: boolean;
  envKey: string;
  policy: string;
  allowedDestinationClasses: string[];
  blockedErrorReason: string;
}>>;

export type OutboundActivityItem = Override<CamelizeKeys<OpenApiItem>, {
  occurredAt: string;
  decision: OutboundDecision;
  destinationClass: string;
  scheme: string;
  hostType: string;
  reason: string;
  correlationId: string;
  localOnlyMode: boolean;
  allowlisted: boolean;
}>;

export type OutboundActivityPage = Override<CamelizeKeys<OpenApiPage>, {
  localOnlyMode: boolean;
  items: OutboundActivityItem[];
  limit: number;
  returned: number;
  maxRetained: number;
}>;

export type OutboundActivityListQuery = { limit?: number };
